whole-tale/ngx-dashboard
# Fix setting GIRDER_API_URL envvar at runtime

## Problem
`GIRDER_API_URL` needs to be set before everything loads, but currently uses an async fetch.

This is problematic because the UI code needs this value when it first loads - setting it after that can lead to failed requests or stale/bad data in the browser.

Fixes #255 

## Approach
Our previous iteration of the dashboard did not have this problem, since it loaded the envs as raw JS instead of a JSON fetch.

Some adjustments to our templates / docker patterns made this work:
* `env.template.js` now produces `env.js` - this JS file can be loaded directly in `index.html`
* Load `env.js` in the `index.html` - this ensures that envs have been loaded before first render or any requests are sent
* Adjust entrypoint / source to use `env.js` instead of `env.json` throughout

## How to Test
1. Checkout this branch locally, rebuild the dashboard
2. Set `GIRDER_API_URL="https://google.com"` and start up the dashboard
3. Navigate to the WholeTale Dashboard
    * You should see failures in the console trying to access `https://google.com/users/me`
4. Shut it down and restart with `GIRDER_API_URL="https://girder.local.wholetale.org/api/v1"`
    * You should see the UI now loads as expected
